File: src/Main/java/com/nishant/algorithms/DataStructures/Trees/Trie/Trie.java
package com.nishant.algorithms.DataStructures.Trees.Trie;

public class Trie {

  private static class TrieNode {
    public TrieNode[] children;
    public int count;

    public TrieNode() {
      children = new TrieNode[26];
    }
  }

  private final TrieNode root;

  public Trie() {
    root = new TrieNode();
  }

  public void insert(String s) {
    TrieNode curr = root;
    char[] chars = s.toCharArray();

    for (char aChar : chars) {
      int index = aChar - 'a';

      if (curr.children[index] == null) {
        curr.children[index] = new TrieNode();
      }

      curr = curr.children[index];
    }

    curr.count++;
  }

  public boolean contains(String s) {
    TrieNode curr = root;
    char[] chars = s.toCharArray();

    for (char aChar : chars) {
      int index = aChar - 'a';

      if (curr.children[index] == null) {
        return false;
      }

      curr = curr.children[index];
    }

    return curr.count > 0;
  }

  public static void main(String[] args) {
    Trie trie = new Trie();
    trie.insert("app");
    trie.insert("apple");
    trie.insert("ap");

    System.out.println(trie.contains("app"));
    System.out.println(trie.contains("appl"));
  }
}

File: src/main/cpp/com/nishant/algorithms/datastructures/trees/trie/trie.cpp
#include <string>
#include <vector>
#include <iostream>

class trie {
private:
	class node {
	public:
		int count;
		std::vector<node*> children;

		node() {
			children = std::vector<node*>(26, nullptr);
		}
	};
public:
	node* root;

	trie() {
		root = new node();
	}

	void insert(std::string string) {
		node* curr = root;

		for(int i = 0; i < string.length(); i ++) {
			int index = string[i] - 'a';

			if(curr->children[index] == nullptr) {
				curr->children[index] = new node();
			}

			curr = curr->children[index];
		}

		curr->count ++;
	}

	bool contains(std::string string) {
		node* curr = root;

		for(int i = 0; i < string.length(); i ++) {
			int index = string[i] - 'a';

			if(curr->children[index] == nullptr) {
				return false;	
			}

			curr = curr->children[index];
		}

		return curr->count > 0;
	}
		
};

int main() {
	trie t;
	t.insert("hi");
	t.insert("hello");

	std::cout << std::boolalpha << t.contains("hi") << std::endl;
	std::cout << std::boolalpha << t.contains("h") << std::endl;
}

